### ISSUE TITLE
About SparseConvTensor()

### ISSUE THREAD

**Issue by Time-Lord12th** [traveller59/spconv#374]
I'm transforming a spconv network to ordinary 3D convolution network, so I can compare their differences.
But I don't understand `SparseConvTensor()` very well, can it be replaced by some torch.nn functions?
Or could you explain the purpose of `SparseConvTensor()`? I didn't find very detailed description.

Thanks.

**Comment by FindDefinition** [comment]
[SparseConvTensor](https://github.com/traveller59/spconv/blob/master/docs/USAGE.md#concept) is features + coordicates.
If you want to convert a network to dense 3d, just
1. ```nn.BatchNorm1d``` to ```nn.BatchNorm3d```
2. ```SparseConv3d``` to ```Conv3d```
3. ```SubMConv3d``` to ```Conv3d``` and make sure stride=1, padding is set to value that produce same shape between input and output
4. ```SparseMaxPool3d``` to ```MaxPool3d```
5. ```SparseInverseConv3d``` to ```Conv3d``` and set parameters according to it's couple
5. ```SparseConvTranspose3d``` to ```ConvTranspose3d``` 

Keep in mind that only SparseConv3d, MaxPool and SparseConvTranspose3d are equivalent to dense conv3d/pool/deconv3d.

**Comment by Time-Lord12th** [comment]
Yeah, I already finished the replacement you mentioned. But the input of dense 3dconv should be Tensor, so I used `.dense()` to convert the output of `SparseConvTensor()` to Tensor. And I trained/tested the network successfully.
Maybe my method seems strange, I wonder if I can replace  `SparseConvTensor()`  directly.

Thanks for your early and detailed reply.

**Comment by Carl12138aka** [comment]
> 

excuse me, did you find out the way to convert SparseConvTensor() directly? Thank you for your reply!

**Comment by Time-Lord12th** [comment]
Oh, I just used `.dense()` to convert the output of `SparseConvTensor()` to Tensor and it worked. Changing the underlying code is difficult to me now :(

**Comment by Carl12138aka** [comment]
> Oh, I just used `.dense()` to convert the output of `SparseConvTensor()` to Tensor and it worked. Changing the underlying code is difficult to me now :(

thank u anyway~ maybe I will try other methods

**Comment by Haerxu** [comment]
> [SparseConvTensor](https://github.com/traveller59/spconv/blob/master/docs/USAGE.md#concept) is features + coordicates. If you want to convert a network to dense 3d, just
> 
> 1. `nn.BatchNorm1d` to `nn.BatchNorm3d`
> 2. `SparseConv3d` to `Conv3d`
> 3. `SubMConv3d` to `Conv3d` and make sure stride=1, padding is set to value that produce same shape between input and output
> 4. `SparseMaxPool3d` to `MaxPool3d`
> 5. `SparseInverseConv3d` to `Conv3d` and set parameters according to it's couple
> 6. `SparseConvTranspose3d` to `ConvTranspose3d`
> 
> Keep in mind that only SparseConv3d, MaxPool and SparseConvTranspose3d are equivalent to dense conv3d/pool/deconv3d. so you can only compare results in macroscopic view, don't compare their value.

Hi,

Why "SparseInverseConv3d to Conv3d"? I thought SparseInverseConv3d is an upsampling function and should corresponds to nn.upsampling?
